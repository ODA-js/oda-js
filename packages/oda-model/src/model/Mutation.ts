import { Map, Record, Set } from 'immutable';

import { IFieldArgInit } from '../interfaces/IFieldArg';
import {
  IMutation,
  IMutationACLStore,
  IMutationACLTransform,
  IMutationInit,
  IMutationStore,
  IMutationTransform,
} from '../interfaces/IMutation';
import { IPackage } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { TransformArgs } from './utils';
import { IPackageContext } from '../contexts/IPackageContext';

// tslint:disable-next-line:variable-name
export const DefaultMutation: IMutationStore = {
  name: null,
  title: null,
  description: null,
  args: null,
  payload: null,
  acl: {
    execute: Set<string>(),
  },
};

// tslint:disable-next-line:variable-name
export const MutationACLTransform: IMutationACLTransform = {
  execute: {
    transform: (input: string[]) => Set<string>(input),
    reverse: (input: Set<string> | string[]) => {
      if (Array.isArray(input)) {
        return input;
      } else {
        return Array.from(input.values()[Symbol.iterator]());
      }
    },
  },
};

// tslint:disable-next-line:variable-name
export const MutationTransform: IMutationTransform = {
  args: TransformArgs(),
  payload: TransformArgs(),
  acl: MutationACLTransform,
};

// tslint:disable-next-line:variable-name
const MutationStorage = Record(DefaultMutation);

export class Mutation extends Persistent<IMutationInit, IMutationStore, IPackageContext> implements IMutation {
  public get modelType(): 'mutation' {
    return 'mutation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get acl(): IMutationACLStore {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get args(): Map<string, IFieldArgInit> {
    return Map(this.store.get('args', null));
  }
  public get payload(): Map<string, IFieldArgInit> {
    return Map(this.store.get('payload', null));
  }

  protected transform(input: Partial<IMutationInit>): Partial<IMutationStore> {
    const result: IMutationStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'args') {
            result.args = MutationTransform.args.transform(input.args, this);
          } else if (f === 'payload') {
            result.payload = MutationTransform.payload.transform(input.payload, this);
          } else if (f === 'acl') {
            for (let facl in input.acl) {
              if (input.acl.hasOwnProperty(facl)) {
                result.acl = {} as any;
                if (facl === 'execute') {
                  result.acl.execute = MutationTransform.acl.execute.transform(input.acl.execute);
                } else {
                  result[facl] = input[f];
                }
              }
            }
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IMutationStore> & Readonly<IMutationStore>): IMutationInit {
    const result: IMutationInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'args') {
            result.args = MutationTransform.args.reverse(input.args);
          } else if (f === 'payload') {
            result.payload = MutationTransform.payload.reverse(input.payload);
          } else if (f === 'acl') {
            const acl = input.acl;
            for (let facl in acl) {
              if (acl.hasOwnProperty(facl)) {
                result.acl = {} as any;
                if (facl === 'execute') {
                  result.acl.execute = MutationTransform.acl.execute.reverse(acl.execute);
                } else {
                  result.acl[facl] = core.acl[facl];
                }
              }
            }
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IMutationInit>, context?: IPackageContext) {
    super();
    if (!context && init && (init.payload || init.args)) {
      throw new Error('context must be provided');
    }
    if (context) {
      this.attach(context);
    }
    this.store = new MutationStorage(this.transform(init));
  }
}
