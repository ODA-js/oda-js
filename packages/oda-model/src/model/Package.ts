import { Map, Record } from 'immutable';

import { IModelContext } from '../contexts/IModelContext';
import { IPackage, IPackageInit, IPackageStore, IPackageTransform } from '../interfaces/IPackage';
import { IPackagedItem, IPackagedItemInit, PackagedItemInit } from '../interfaces/IPackagedItem';
import { Persistent } from './Persistent';

// tslint:disable-next-line:variable-name
export const DefaultPackage: IPackageStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  abstract: null,
  items: null,
};

// tslint:disable-next-line:variable-name
export const PackageTransform: IPackageTransform = {
  items: {
    transform: (input: PackagedItemInit[], pkg: Package) => {
      if (input) {
        return Map<string, IPackagedItem>(input.map(p => {
          if (typeof p === 'string') {
            if (pkg.context.model.defaultPackage.items.has(p)) {
              return [p, {
                ...pkg.context.model.defaultPackage.items.get(p),
                package: pkg,
              }];
            } else {
              throw Error('item does not exists');
            }
          } else {
            return [p.name, {
              ...p,
              package: pkg,
            }];
          }
        },
        ) as [string, IPackagedItem][]);
      } else {
        return null;
      }
    },
    reverse: (input?: Map<string, IPackagedItem>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(i => i.toJS() as IPackagedItemInit);
      } else {
        return null;
      }
    },
  },
};

// tslint:disable-next-line:variable-name
const PackageStorage = Record(DefaultPackage);

export class Package extends Persistent<IPackageInit, IPackageStore, IModelContext> implements IPackage {
  public get modelType(): 'package' {
    return 'package';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get abstract(): boolean {
    return this.store.get('abstract', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl(): number {
    return this.store.get('acl', null);
  }
  public get items(): Map<string, IPackagedItem> {
    return this.store.get('items', null);
  }

  protected transform(input: Partial<IPackageInit>): IPackageStore {
    const result: IPackageStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.transform(input.items, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IPackageStore> & Readonly<IPackageStore>): IPackageInit {
    const result: IPackageInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.reverse(input.items);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: Partial<IPackageInit>, context: IModelContext) {
    super();
    if (!context && init.items) {
      throw new Error('context must be provided');
    }
    this.attach(context);
    this.store = new PackageStorage(this.transform(init));
    this.init = new (Record<Partial<IPackageInit>>(init))();
  }
}
