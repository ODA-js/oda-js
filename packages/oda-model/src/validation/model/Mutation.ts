import { Record } from 'immutable';
import { Map, Set } from 'immutable';

import { IFieldArgs } from '../interfaces/IField';
import {
  IMutation,
  IMutationACLStore,
  IMutationProps,
  IMutationPropsStore,
  IMutationTransform,
} from '../interfaces/IMutation';
import { Persistent } from './Persistent';
import { transformMap, transformSet } from './utils';

// tslint:disable-next-line:variable-name
export const DefaultMutation: IMutationPropsStore = {
  name: null,
  title: null,
  description: null,
  args: null,
  payload: null,
  acl: {
    execute: Set<string>(),
  },
};

// tslint:disable-next-line:variable-name
export const MutationTransform: IMutationTransform = {
  args: transformMap<IFieldArgs>(),
  payload: transformMap<IFieldArgs>(),
  acl: {
    execute: transformSet<string>(),
  },
};

// tslint:disable-next-line:variable-name
const MutationStorage = Record(DefaultMutation);

export class Mutation extends Persistent<IMutationProps, IMutationPropsStore> implements IMutation {
  public get modelType(): 'mutation' {
    return 'mutation';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get acl(): IMutationACLStore {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get args(): Map<string, IFieldArgs> {
    return this.store.get('args', null);
  }
  public get payload(): Map<string, IFieldArgs> {
    return this.store.get('payload', null);
  }

  protected transform(input: IMutationProps): IMutationPropsStore {
    return {
      ...input,
      args: MutationTransform.args.transform(input.payload),
      payload: MutationTransform.args.transform(input.payload),
      acl: {
        execute: MutationTransform.acl.execute.transform(input.acl.execute),
      },
    };
  }

  protected reverse(input: IMutationPropsStore): IMutationProps {
    return {
      ...input,
      args: MutationTransform.args.reverse(input.payload),
      payload: MutationTransform.args.reverse(input.payload),
      acl: {
        execute: MutationTransform.acl.execute.reverse(input.acl.execute),
      },
    };
  }

  constructor(init: IMutationProps) {
    super();
    this.store = new MutationStorage(this.transform(init));
    this.init = new (Record<IMutationProps>(init))();
  }
}
