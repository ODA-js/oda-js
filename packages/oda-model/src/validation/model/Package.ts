import { Record } from 'immutable';
import { Map } from 'immutable';

import { IModel } from '../interfaces/IModel';
import { IModelType } from '../interfaces/IModelType';
import { IPackage, IPackageInit, IPackageStore, IPackageTransform } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { transformMap } from './utils';
import {IPackagedItem } from '../interfaces/IPackagedItem';

// tslint:disable-next-line:variable-name
export const DefaultPackage: IPackageStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  abstract: null,
  items: null,
  model: null,
};

// tslint:disable-next-line:variable-name
export const PackageTransform: IPackageTransform = {
  items: transformMap<IPackagedItem>(),
};

// tslint:disable-next-line:variable-name
const PackageStorage = Record(DefaultPackage);

export class Package extends Persistent<IPackageInit, IPackageStore> implements IPackage {
  public get modelType(): 'package' {
    return 'package';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get abstract(): boolean {
    return this.store.get('abstract', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl(): number {
    return this.store.get('acl', null);
  }
  public get items(): Map<string, IPackagedItem> {
    return this.store.get('items', null);
  }
  public get model(): IModel {
    return this.store.get('model', null);
  }

  protected transform(input: IPackageInit): IPackageStore {
    const result: IPackageStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.transform(input.items);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: IPackageStore): IPackageInit {
    const result: IPackageInit = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'items') {
            result.items = PackageTransform.items.reverse(input.items);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init: IPackageInit) {
    super();
    this.store = new PackageStorage(this.transform(init));
    this.init = new (Record<IPackageInit>(init))();
  }
}
