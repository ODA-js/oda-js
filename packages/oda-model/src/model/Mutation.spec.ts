import { Mutation } from './Mutation';
import { Package } from './Package';

describe('Mutaion', () => {
  let mutation: Mutation;

  beforeAll(() => {
    expect(() => new Mutation()).not.toThrow();
  });

  beforeEach(() => {
    mutation = new Mutation();
  });

  it('create empty', () => {
    expect(mutation).toMatchSnapshot();
    expect(mutation.toJS()).toMatchSnapshot();
  });

  it('updates strings', () => {
    expect(() => mutation.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();
    expect(() => mutation.updateWith({
      name: 'cool!',
    })).not.toThrow();
    expect(mutation.name).toBe('cool!');
    expect(mutation.toJS()).toMatchSnapshot();
  });

  it('updates with null or undefined', () => {
    expect(() => mutation.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();
    expect(() => mutation.updateWith({
      description: null,
      title: undefined,
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();
  });

  it('updates acl', () => {
    expect(() => mutation.updateWith({
      acl: {
        execute: ['admin', 'public', 'public'],
      },
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();

    expect(() => mutation.updateWith({
      acl: {
        execute: ['system'],
      },
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();
  });

  it('updates input args', () => {
    expect(() => mutation.updateWith({
      args: [],
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();

    expect(() => mutation.updateWith({
      args: [
        { name: '1', type: 'string', defaultValue: 'not empty', required: true }],
    })).not.toThrow();
    let arg;
    expect(() => arg = mutation.args.get('1')).not.toThrow();
    expect(arg.toJS()).toMatchSnapshot();

    expect(() => mutation.updateWith({
      args: [
        { name: '1', type: 'number' }]
      ,
    })).not.toThrow();
    expect(() => arg = mutation.args.get('1')).not.toThrow();
    expect(arg.toJS()).toMatchSnapshot();
  });

  it('updates payload', () => {
    expect(() => mutation.updateWith({
      payload: [],
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();

    expect(() => mutation.updateWith({
      payload: [
        { name: '1', type: 'string', defaultValue: 'not empty', required: true }],
    })).not.toThrow();
    let arg;
    expect(mutation.payload.has('1'));
    expect(() => arg = mutation.payload.get('1')).not.toThrow();

    expect(arg.toJS()).toMatchSnapshot();
    expect(() => mutation.updateWith({
      payload: [
        { name: '1', type: 'number' }]
      ,
    })).not.toThrow();
    expect(() => arg = mutation.payload.get('1')).not.toThrow();
    expect(arg.toJS()).toMatchSnapshot();
  });

  it('toJS with dupes', () => {
    expect(() => mutation.updateWith({
      name: 'cool',
      description: 'very cool',
      title: 'very cool title',
      acl: {
        execute: ['admin', 'public', 'public'],
      },
      args: [
        { name: 'i1', type: 'number', defaultValue: '100', required: true },
        { name: 'i1', type: 'string', defaultValue: 'not empty', required: true },
        { name: 'i2', type: 'number', defaultValue: '1', required: false },
      ],
      payload: [
        { name: 'p1', type: 'string', defaultValue: 'not empty', required: true },
        { name: 'p1', type: 'number', defaultValue: '10', required: true },
        { name: 'p2', type: 'number', defaultValue: '1', required: false },
      ],
    })).not.toThrow();
    expect(mutation.toJS()).toMatchSnapshot();
  });

});
