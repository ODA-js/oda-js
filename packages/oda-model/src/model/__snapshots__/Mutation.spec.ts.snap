// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutaion create empty 1`] = `
Mutation {
  "store": Immutable.Record {
    "name": null,
    "title": null,
    "description": null,
    "args": null,
    "payload": null,
    "acl": Object {
      "execute": Immutable.Set [],
    },
  },
}
`;

exports[`Mutaion create empty 2`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
}
`;

exports[`Mutaion toJS with dupes 1`] = `
Object {
  "acl": Object {
    "execute": Array [
      "admin",
      "public",
    ],
  },
  "args": Array [
    Object {
      "defaultValue": "not empty",
      "name": "i1",
      "required": true,
      "type": "string",
    },
    Object {
      "defaultValue": "1",
      "name": "i2",
      "required": false,
      "type": "number",
    },
  ],
  "description": "very cool",
  "name": "cool",
  "payload": Array [
    Object {
      "defaultValue": "10",
      "name": "p1",
      "required": true,
      "type": "number",
    },
    Object {
      "defaultValue": "1",
      "name": "p2",
      "required": false,
      "type": "number",
    },
  ],
  "title": "very cool title",
}
`;

exports[`Mutaion updates acl 1`] = `
Object {
  "acl": Object {
    "execute": Array [
      "admin",
      "public",
    ],
  },
}
`;

exports[`Mutaion updates acl 2`] = `
Object {
  "acl": Object {
    "execute": Array [
      "admin",
      "public",
      "system",
    ],
  },
}
`;

exports[`Mutaion updates input args 1`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "args": Array [],
}
`;

exports[`Mutaion updates input args 2`] = `
Object {
  "defaultValue": "not empty",
  "name": "1",
  "required": true,
  "type": "string",
}
`;

exports[`Mutaion updates input args 3`] = `
Object {
  "name": "1",
  "type": "number",
}
`;

exports[`Mutaion updates payload 1`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "payload": Array [],
}
`;

exports[`Mutaion updates payload 2`] = `
Object {
  "defaultValue": "not empty",
  "name": "1",
  "required": true,
  "type": "string",
}
`;

exports[`Mutaion updates payload 3`] = `
Object {
  "name": "1",
  "type": "number",
}
`;

exports[`Mutaion updates strings 1`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Mutaion updates strings 2`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "description": "very cool",
  "name": "cool!",
  "title": "very cool title",
}
`;

exports[`Mutaion updates with null or undefined 1`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "description": "very cool",
  "name": "cool",
  "title": "very cool title",
}
`;

exports[`Mutaion updates with null or undefined 2`] = `
Object {
  "acl": Object {
    "execute": Array [],
  },
  "name": "cool",
}
`;
